src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+def _series_nan(length: int) -> "pd.Series[float]":
+    """Useful as a substitute for a non-existent column."""
+    return pd.Series(np.full(length, float("nan")))
+
+
+DataFrameOrSeries: TypeAlias = Union[pd.DataFrame, "pd.Series[Any]"]
+
+
+@dataclass(frozen=True)
+class Column:
+    """Extracts value from pd.Series or series from pd.DataFrame. If not
+    found, returns NaN or a series of NaNs, respectively.
+    """
+
+    name: str = ""
+    is_dummy: bool = False
+
+    def __post_init__(self) -> None:
+        if not self.name:
+            object.__setattr__(self, "is_dummy", True)
+            object.__setattr__(self, "name", _rand_str())
+
+    @overload
+    def __call__(self, data: pd.DataFrame) -> "pd.Series[Any]":
+        ...
+
+    @overload
+    def __call__(self, data: "pd.Series[Any]") -> Any:
+        ...
+
+    def __call__(self, data: Any) -> Any:
+        """Extracts series from dataframe, or value from series,
+        returning NaN if not found.
+        """
+        if isinstance(data, pd.DataFrame):
+            try:
+                return data.loc[:, self.name]
+            except KeyError:
+                return _series_nan(len(data)).set_axis(data.index)
+        if isinstance(data, pd.Series):
+            try:
+                return data.at[self.name]
+            except KeyError:
+                return float("nan")
+        raise ValueError("invalid data: %s" % repr(data))
+
+    def __iter__(self) -> Iterator[str]:
+        """This is to partake in the iteration of column names by a
+        larger data structure of which this object is a member.
+        """
+        yield self.name
+
+    def __str__(self) -> str:
+        return self.name
+
+
+class DataType(IntEnum):
+    UNKNOWN = auto()
+    DISCRETE = auto()
+    CONTINUOUS = auto()
+
+
+UNKNOWN = DataType.UNKNOWN
+DISCRETE = DataType.DISCRETE
+CONTINUOUS = DataType.CONTINUOUS
+
+
+@dataclass(frozen=True)
+class Dimension(Column, ABC):
+    """All Dimensions are Columns (i.e. dataframe column) but not all Columns
+    are Dimensions, e.g. some are just ancillary, like a column of URLs.
+    """
+
+    role: DimensionRole = UNASSIGNED
+    """The dimension's role in the Model, e.g. FEATURE, TIME, or PREDICTION_ID.
+    Must not be unassigned at initialization."""
+    data_type: DataType = UNKNOWN
+    """Denotes whether the dimension values are continuous or discrete, useful
+    for binning."""
+    _model: Optional["ProxyType[Model]"] = field(repr=False, default=None)
+    """Holds a weak reference to the parent model, if any. This is the channel
+    for talking to actual data, so the model can also help out if needed."""
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        if self.role is UNASSIGNED:
+            # This is to let the superclass be initialized without arguments.
+            # But we really want the role to be specified for a Dimension.
+            raise ValueError("role must be assigned")
```

Comment:
any reason why this is done as a post-init runtime error rather than rely on types and the constructor?